packages/puppeteer-core/src/common/bidi/HTTPRequest.ts

```diff
@@ -0,0 +1,115 @@
+/**
+ * Copyright 2020 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import * as Bidi from 'chromium-bidi/lib/cjs/protocol/protocol.js';
+
+import {Frame} from '../../api/Frame.js';
+import {
+  HTTPRequest as BaseHTTPRequest,
+  ResourceType,
+} from '../../api/HTTPRequest.js';
+
+import {HTTPResponse} from './HTTPResponse.js';
+
+/**
+ * @internal
+ */
+export class HTTPRequest extends BaseHTTPRequest {
+  override _response: HTTPResponse | null = null;
+  override _redirectChain: HTTPRequest[];
+  _navigation: string | null;
+
+  #url: string;
+  #resourceType: ResourceType;
+
+  #method: string;
+  #postData?: string;
+  #headers: Record<string, string> = {};
+  #initiator: Bidi.Network.Initiator;
+  #frame: Frame | null;
+
+  constructor(
+    event: Bidi.Network.BeforeRequestSentParams,
+    frame: Frame | null,
+    redirectChain: HTTPRequest[]
+  ) {
+    super();
+
+    this.#url = event.request.url;
+    this.#resourceType = event.initiator.type.toLowerCase() as ResourceType;
+    this.#method = event.request.method;
+    this.#postData = undefined;
+    this.#initiator = event.initiator;
+    this.#frame = frame;
+
+    this._requestId = event.request.request;
+    this._redirectChain = redirectChain ?? [];
+    this._navigation = event.navigation;
+
+    for (const {name, value} of event.request.headers) {
+      // TODO: How to handle Binary Headers
```

Comment:
do we support binary headers? is it a bidi feature or Puppeteer's?

Reply:
It's a BiDi feature - https://w3c.github.io/webdriver-bidi/#type-network-Header will leave the link the the comment